Include/internal/pycore_object.h

```diff
@@ -29,6 +29,9 @@ _PyType_HasFeature(PyTypeObject *type, unsigned long feature) {
 
 extern void _PyType_InitCache(PyInterpreterState *interp);
 
+PyAPI_FUNC(PyObject *) _PyType_DisabledNew(PyTypeObject *type, PyObject *args,
```

Comment:
Maybe make `PyType_FromSpec()` supporting `{Py_tp_new, NULL}`?

Reply:
+1

IMHO this solution is good enough for beta1. We can improve the code for beta2.

Reply:
> Maybe make PyType_FromSpec() supporting {Py_tp_new, NULL}?

type_new() and PyType_Ready() code for slots and methods is very complex. Rather than using NULL as a special marker, I would prefer to use a type flag. IMO a flag is more reliable, since tp_new can be overriden in many different non-obvious ways.

I created PR #25721 which adds Py_TPFLAGS_DISABLE_NEW flag. When the flag is set, tp_new is ignored and TypeError is raised. It avoids the need to include an internal header file, it can be used outside CPython code base, and it simplifies the code (see the PR).